sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,807 @@
+        for different data samples, we use private copies of the new centroid array in
+        global memory. Those copies are meant to be re-reduced afterwards.
+
+        The distance is the euclidean distance. Note that it is not necessary to
+        compute the exact value to find the closest centroid. Indeed, minimizing
+            |x-c|^2 = |x|^2 - 2<x.c> + |c|^2
+        over c for a given x amounts to minimizing
+            (1/2)c^2 - xc .
+        Moreover the value (1/2)c^2 has been pre-computed in the array
+        centroids_half_l2_norm to reduce the overall number of floating point
+        operations in the kernel.
+        """
+        sample_idx = dpex.get_global_id(0)
+        local_work_id = dpex.get_local_id(0)
+
+        # This array in shared memory is used as a sliding array over values of
+        # current_centroids_t. During each iteration in the inner loop, a new one is
+        # loaded and used by all work items in the work group to compute partial
+        # results. The array slides over the features in the outer loop, and over the
+        # samples in the inner loop.
+        centroids_window = dpex.local.array(shape=centroids_window_shape, dtype=dtype)
+
+        # This array in shared memory is used as a sliding array over the centroids.
+        # It contains values of centroids_half_l2_norm for each centroid in the sliding
+        # centroids_window array. It is updated once per iteration in the outer loop.
+        centroids_window_half_l2_norm = dpex.local.array(
+            shape=window_n_centroids, dtype=dtype
+        )
+
+        # In the inner loop each work item accumulates in private memory the
+        # dot product of the sample at the sample_idx relatively to each centroid
+        # in the window.
+        partial_dot_products = dpex.private.array(shape=window_n_centroids, dtype=dtype)
+
+        first_centroid_idx = 0
+
+        # The two variables that are initialized here will contain the result we seek,
+        # i.e, at the end of the outer loop, it will be equal to the closest centroid
+        # to the current sample and the corresponding pseudo inertia.
+        min_idx = 0
+        min_sample_pseudo_inertia = inf
+
+        # Those variables are used in the inner loop during loading of the window of
+        # centroids
+        window_loading_centroid_idx = local_work_id % window_n_centroids
+        window_loading_feature_offset = local_work_id // window_n_centroids
+
+        # STEP 1: compute the closest centroid
+        # Outer loop: iterate on successive windows of size window_n_centroids that
+        # cover all centroids in current_centroids_t
+        for _0 in range(n_windows_per_feature):
+
+            # initialize the partial pseudo inertia for each of the window_n_centroids
+            # centroids in the window
+            for i in range(window_n_centroids):
+                partial_dot_products[i] = zero
```

Comment:
```suggestion
            # Initialize dot products arrays used for computing the 
            # partial pseudo inertia for each of the `window_n_centroids`
            # centroids in the window
            for i in range(window_n_centroids):
                dot_products[i] = zero
```

alternatively to be verbose:

```suggestion
            # Initialize dot products arrays used for computing the 
            # partial pseudo inertia for each of the `window_n_centroids`
            # centroids in the window
            for i in range(window_n_centroids):
                partial_inertia_dot_products[i] = zero
```